---
title: "Association analysis"
output: html_notebook
---

# Running the analysis

We will start by running an association analysis using DESeq2. In short we will extract our data subsets (particular sample site
at a particular time) and get the association with basic plant growth indicators and the inoculum.

```{r, warning = FALSE}
library(mbtools)
theme_set(theme_classic())

ps <- readRDS("data/taxonomy.rds")
otu_table(ps)[otu_table(ps) < 1] <- 0
sdata <- fread("data/plant_data_clean.csv")
sdata <- sdata[order(date)][sample == "Growth Bed", .(no_leaves=no_leaves_avg[.N], 
                                                      height=height_avg[.N], 
                                                      root_length=root_length_avg[.N]), 
                                                   by=tank]
old <- fread("data/barcodes.csv")
old[, inoculum := factor(inoculum, levels = c("EIT", "CIT"))]
old[, extraction_kit := factor(extraction_kit)]
sdata <- as.data.frame(sdata[old, on = "tank"])
# Read the greenhouse data
greenhouse <- fread("data/greenhouse.csv")
greenhouse[, date := as.Date(date, format = "%m/%d/%y")]
greenhouse[, time := difftime(date, date[1], units = "days")]
greenhouse[, inoculum := "aquaponics system"]
greenhouse[tank %in% 1:3, inoculum := "CIT"]
greenhouse[, tank := factor(tank)]
greenhouse <- greenhouse[order(-time, tank)]
greenhouse <- greenhouse[, .(nitrate = sum(as.numeric(nitrate)), ammonia = sum(ammonia)), by="tank"]
sdata <- merge(sdata, greenhouse, by = "tank", all = TRUE)
sdata$isb_id[sdata$isb_id == "T0RW"] <- "T0R"
sample_names(ps)[sample_names(ps) == "T0RW"] <- "T0R"
rownames(sdata) <- sdata$isb_id
sample_data(ps) <- sdata
```


```{r}
subsets <- list(c("biofilter", "T2"), c("fish feces", "T0"), c("root", "T2"))
vars <- c("no_leaves", "height", "root_length", "inoculum", "nitrate", "ammonia", "extraction_kit", "dead_fish")
tests <- list()
for (s in subsets) {
    psub <- subset_samples(ps, compartment == s[1] & sample_set == s[2])
    res <- association(psub, variables=vars, min_count = 10, in_samples=0.1)
    res[, "site" := s[1]]
    tests[[s[1]]] <- res
}

tests <- rbindlist(tests)
fwrite(tests, "data/associations.csv")
```

Now let's readjust the p values and check the histogram.

```{r}
ggplot(tests, aes(x=pvalue)) + geom_histogram(bins=20) + facet_wrap(~ site)
```

We will also track a list of genera with abundances larger than our cutoff in at least 2 samples:

```{r}
counts <- taxa_count(ps, lev = "Genus") %>% as.matrix()
genera <- colnames(counts)[apply(counts, 2, function(x) sum(x > 300)) > 1]
genera <- genera[!grepl("uncultured|metagenome", genera)]
```

First let's have a look which genera associate with our confounders.

```{r, fig.width=8, fig.height=2}
confounders <- c("nitrogen", "ammonia", "nitrate", "dead_fish", "extraction_kit")
ggplot(tests[padj < 0.05 & genus %chin% genera & variable %in% confounders],
       aes(x = log2FoldChange, y = reorder(genus, -log2FoldChange),
           color = baseMean)) +
    geom_vline(xintercept = 0, lty = "dashed", color = "black") +
    geom_point(size=2) +
    geom_errorbarh(aes(xmin = log2FoldChange - lfcSE,
                    xmax = log2FoldChange + lfcSE), height = 0.5) +
    facet_grid(site ~ variable, scales = "free_y", space = "free_y") + 
    scale_color_gradient(trans="log10") +
    labs(x="log fold change", y="", color="abundance") + theme_minimal()
ggsave(sprintf("figures/confounders.svg"), width = 8, height = 2)
```

Let's visuazlize the effect sizes and significant tests.

```{r, fig.width=8, fig.height=3}
ggplot(tests[padj < 0.05 & genus %chin% genera & !variable %in% confounders],
       aes(x = log2FoldChange, y = reorder(genus, -log2FoldChange),
           color = baseMean)) +
    geom_vline(xintercept = 0, lty = "dashed", color = "black") +
    geom_point(size=2) +
    geom_errorbarh(aes(xmin = log2FoldChange - lfcSE,
                    xmax = log2FoldChange + lfcSE), height = 0.5) +
    facet_grid(site ~ variable, scales = "free_y", space = "free_y") + 
    scale_color_gradient(trans="log10") +
    labs(x="log fold change", y="", color="abundance") + theme_minimal()
ggsave(sprintf("figures/lfcs.svg"), width = 8, height = 3)
```


## Temporal development

Let's have a look for the genera that are specifically targeting the root. 
For that we wil first assemble the corresponding data set:

```{r}
counts <- taxa_count(ps, zeros = TRUE)
sdata <- as.data.table(sdata)
counts <- counts[sdata, on = c(sample = "isb_id")]
counts[, relative := reads / sum(reads), by = "sample"]

genera <- c("Pseudomonas", "Aeromonas", "Serratia", "Rahnella")
counts <- counts[taxa %in% genera & compartment %in% c("biofilter", "root", "inoculum", "rock wool cubes")]
map <- rep(c("CIT", "EIT"), each = 3)
initial <- counts[is.na(tank)]
initial <- lapply(1:6, function(i) {
  co <- copy(initial) 
  co$tank <- i
  co$inoculum <- map[i]
  co$compartment[co$compartment == "inoculum"] <- "biofilter"
  return(co)
})
initial[[7]] <- counts[!is.na(tank)]
counts <- rbindlist(initial)
counts[, inoculum := factor(inoculum, levels = c("CIT", "EIT"))]


```

Now we can visualize the temporal changes:

```{r, fig.width = 8, fig.height = 3}
ggplot(counts, aes(x = sample_set, y = relative, col = inoculum, group = interaction(tank, inoculum))) +
  geom_point() + geom_line() +
  facet_grid(compartment ~ taxa, scales = "free_y") +
  scale_y_continuous(labels = scales::percent) +
  labs(x = "time point", y = "relative abundance") + theme_minimal()
ggsave("figures/pathogens_time.svg", width = 8, height = 3)
```