---
title: "General plots"
output: html_notebook
---

We will start by loading the data set:

```{r}
library(mbtools)
library(ggplot2)
theme_set(theme_classic())

ps <- readRDS("data/taxonomy.rds")
sdata <- read.csv("data/barcodes.csv", stringsAsFactors = FALSE)
rownames(sdata) <- sdata$isb_id
sdata$isb_id[sdata$isb_id == "T0RW"] <- "T0R"
sample_data(ps) <- sdata
sample_names(ps)[sample_names(ps) == "T0RW"] <- "T0R"
ps
```

Taxa plots

```{r, fig.width = 8, fig.height = 4}
phyla <- plot_taxa(ps, level = "phylum", only_data = TRUE)
phyla <- as.data.table(sdata)[phyla, on = c(isb_id = "sample")]
ggplot(phyla, aes(x = isb_id, y = reads, fill = taxa)) + 
  geom_bar(stat = "identity", position = "stack", size = 0.1, color = "white") +
  facet_wrap(~ inoculum, scales = "free_x") + 
  scale_y_continuous(expand = c(0, 0.01), labels = scales::percent) +
  scale_fill_brewer(palette = "Paired", direction = -1) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  labs(x = "", y = "relative abundance", fill = "")
ggsave("figures/phyla.png", width = 8, height = 4, dpi = 300)
```

## Ordination

```{r, fig.width=5, fig.height=3.5}
inoc_cols <- c(scales::hue_pal()(2), "gray")

ord <- ordinate(rarefy_even_depth(ps, 3000), method="PCoA")
plot_ordination(ps, ord, shape="compartment", color="inoculum") + 
  stat_ellipse(aes(group = compartment, lty = compartment), type = "t") +
  scale_color_manual(values = inoc_cols) +
  guides(lty = FALSE)
ggsave("figures/pcoa.svg", width = 5, height = 3.5)
```

And the corresponding PERMANOVA:

```{r}
library(vegan)

counts <- taxa_count(rarefy_even_depth(ps, 3000), NA)[reads >= 1] %>% as.matrix()
perm <- adonis(counts ~ compartment*inoculum, data = sdata)
perm
```

### Dynamics

```{r}
biofilters <- subset_samples(ps, compartment == "biofilter" | isb_id %in% c("T0B", "T0Comm"))
ord <- ordinate(rarefy_even_depth(biofilters), method="PCoA")
df <- plot_ordination(biofilters, ord, color = "inoculum", justDF = TRUE)

ggplot(df, aes(x=Axis.1, y=Axis.2, color=inoculum)) + geom_point(size=3, alpha=0.5) + 
  geom_line(aes(group=tank), data=function(x) dplyr::filter(x, compartment == "biofilter"), 
            arrow=arrow(ends = "first", type = "closed", length = unit(0.2, "cm"))) + 
  geom_text(aes(label=isb_id), nudge_y=-0.02, size=3) +
  theme_bw() + labs(x=sprintf("Axis.1 [%.1f%%]", ord$values[1, 2] * 100), y=sprintf("Axis.2 [%.1f%%]", ord$values[2, 2] * 100))
ggsave("figures/dynamics.pdf")
```
## Richness

```{r, fig.width=4.5, fig.height=3}
alpha <- cbind(
  estimate_richness(rarefy_even_depth(ps, 3000))[sample_names(ps), ], 
  sample_data(ps))

ggplot(alpha, aes(x = sample_set, y = Shannon, color = inoculum, 
                        shape = compartment, group = inoculum)) +        
        geom_point(position = position_dodge(width=0.5)) +
        scale_color_manual(values = inoc_cols) +
  labs(x = "time point", y = "alpha diversity [Shannon index]")
ggsave("figures/diversity.svg", width = 4.5, height = 3)
```

Corresponding tests:

```{r}
alpha <- as.data.table(alpha)
alpha[compartment == "biofilter" & sample_set == "T1", t.test(Shannon ~ inoculum)]
```

and for T2:

```{r}
alpha[compartment == "biofilter" & sample_set == "T2", t.test(Shannon ~ inoculum)]
```

## Heatmaps

```{r, fig.height=9, fig.width=8}
library(pheatmap)

counts <- as.matrix(taxa_count(ps))
counts <- counts[, apply(counts, 2, function(x) sum(x > 300) > 1)]
counts <- counts[, !grepl("uncultured|metagenome", colnames(counts))]
labels <- sdata[, c("compartment", "sample_set", "inoculum")]
names(labels)[2] <- "time point"
label_cols <- list(
  compartment = c(inoculum = "white", `fish feces` = "lightsteelblue",
                  biofilter = "dodgerblue", root = "darkblue"),
  `time point` = c(T0 = "white", T1 = "gray", T2 = "black"),
  inoculum = c(CIT = inoc_cols[2], EIT = inoc_cols[1], 
               `rock wool cubes` = inoc_cols[3])
)

cols <- viridis::viridis(128) 
pheatmap(log10(t(counts) + 1), color = cols, annotation_col = labels,
         width = 8, height = 9, filename = "figures/heatmap.pdf",
         annotation_colors = label_cols)
pheatmap(log10(t(counts) + 1), color = cols, annotation_col = labels,
         annotation_colors = label_cols)
```

Most abundant:

```{r}
counts <- taxa_count(ps, "Genus")[sdata, on = c(sample = "isb_id")]

counts <- counts[, .(reads = mean(reads)), by = c("taxa", "compartment")]
counts[order(-reads), .(taxa = taxa[1:5], reads = reads[1:5]), by = "compartment"]
```

Heatmap for Nitrifiers:

```{r, fig.height=2.1, fig.width=8}
library(pheatmap)

counts <- as.matrix(taxa_count(ps, zeros = TRUE)[grepl("Nitro", taxa)])
pheatmap(log10(t(counts) + 1))
```

As percentages:

```{r, fig.height=3, fig.width=8}
all_counts <- as.matrix(taxa_count(ps, zeros = TRUE))
pheatmap(t(counts / rowSums(all_counts)) * 100)
```